File: mobile/lib/services/audio_service.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:just_audio/just_audio.dart';

class AudioService {
  factory AudioService() => _instance;
  AudioService._internal();
  static final AudioService _instance = AudioService._internal();

  final AudioPlayer _musicPlayer = AudioPlayer();
  final AudioPlayer _sfxPlayer = AudioPlayer();

  bool _isMusicEnabled = false;
  bool _isSfxEnabled = true;
  bool _isHostControlled = false;
  String _equippedMusic = 'music2.mp3'; // Default music

  final ValueNotifier<bool> musicEnabledNotifier = ValueNotifier<bool>(false);
  final ValueNotifier<bool> sfxEnabledNotifier = ValueNotifier<bool>(true);
  final ValueNotifier<bool> isHostControlledNotifier = ValueNotifier<bool>(
    false,
  );
  final ValueNotifier<String> equippedMusicNotifier = ValueNotifier<String>(
    'music2.mp3',
  );

  bool get isMusicEnabled => _isMusicEnabled;
  bool get isSfxEnabled => _isSfxEnabled;

  bool get musicEnabled => _isMusicEnabled;
  // ignore: unnecessary_getters_setters : dont worry
  bool get sfxEnabled => _isSfxEnabled;
  String get equippedMusic => _equippedMusic;

  void setEquippedMusic(String filename) {
    _equippedMusic = filename;
    equippedMusicNotifier.value = filename;
    // If music is currently playing, switch to the new track
    if (_isMusicEnabled &&
        _musicPlayer.playerState.processingState != ProcessingState.idle) {
      playBackgroundMusic(filename);
    }
  }

  Future<void> playBackgroundMusic(String filename) async {
    if (!_isMusicEnabled) return;
    try {
      await _musicPlayer.setLoopMode(LoopMode.one);
      await _musicPlayer.setVolume(0.5);
      await _musicPlayer.setAsset('lib/assets/sounds/$filename');
      await _musicPlayer.play();
      debugPrint('Successfully playing background music: $filename');
      debugPrint('Player state: ${_musicPlayer.playerState}');
      debugPrint('Player volume: ${_musicPlayer.volume}');
    } on Exception catch (e) {
      debugPrint('Error playing background music: $e');
    }
  }

  Future<void> stopMusic() async {
    _isMusicEnabled = false;
    musicEnabledNotifier.value = false;
    await _musicPlayer.stop();
  }

  Future<void> pauseMusic() async {
    await _musicPlayer.pause();
  }

  Future<void> resumeMusic() async {
    await _musicPlayer.play();
  }

  Future<void> playSfx(String filename) async {
    if (!_isSfxEnabled) return;
    try {
      unawaited(
        _sfxPlayer.setAsset('lib/assets/sounds/$filename').then((_) {
          _sfxPlayer.play();
        }),
      );
    } on Exception catch (e) {
      debugPrint('Error playing sound effect: $e');
    }
  }

  set musicEnabled(bool enabled) {
    _isMusicEnabled = enabled;
    musicEnabledNotifier.value = enabled;
    if (enabled) {
      if (_musicPlayer.playerState.processingState == ProcessingState.idle) {
        playBackgroundMusic(_equippedMusic);
      } else {
        resumeMusic();
      }
    } else {
      pauseMusic();
    }
  }

  set sfxEnabled(bool enabled) {
    _isSfxEnabled = enabled;
    sfxEnabledNotifier.value = enabled;
  }

  void setHostControlledSettings({
    required bool musicEnabled,
    required bool sfxEnabled,
  }) {
    _isHostControlled = true;
    isHostControlledNotifier.value = true;
    this.musicEnabled = musicEnabled;
    this.sfxEnabled = sfxEnabled;
  }

  void clearHostControl() {
    _isHostControlled = false;
    isHostControlledNotifier.value = false;
  }

  bool get isHostControlled => _isHostControlled;

  void dispose() {
    _musicPlayer.dispose();
    _sfxPlayer.dispose();
    musicEnabledNotifier.dispose();
    sfxEnabledNotifier.dispose();
    isHostControlledNotifier.dispose();
    equippedMusicNotifier.dispose();
  }
}

File: client/src/app/services/audio/audio.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class AudioService {
    private musicPlayer: HTMLAudioElement | null = null;
    private _equippedMusic = 'music2.mp3';
    private currentlyLoadedMusic: string | null = null;
    private readonly isMusicEnabledSubject = new BehaviorSubject<boolean>(false);
    private readonly areSoundEffectsEnabledSubject = new BehaviorSubject<boolean>(true);
    private readonly isHostControlledSubject = new BehaviorSubject<boolean>(false);
    public musicEnabled$ = this.isMusicEnabledSubject.asObservable();
    public soundEffectsEnabled$ = this.areSoundEffectsEnabledSubject.asObservable();
    public isHostControlled$ = this.isHostControlledSubject.asObservable();

    get isMusicEnabled(): boolean {
        return this.isMusicEnabledSubject.value;
    }

    set musicEnabled(enabled: boolean) {
        this.isMusicEnabledSubject.next(enabled);
        if (enabled) {
            if (this.musicPlayer) {
                this.resumeMusic();
            } else {
                this.playBackgroundMusic(this._equippedMusic);
            }
        } else if (this.musicPlayer) {
            this.pauseMusic();
        }
    }

    get areSoundEffectsEnabled(): boolean {
        return this.areSoundEffectsEnabledSubject.value;
    }

    set areSoundEffectsEnabled(enabled: boolean) {
        this.areSoundEffectsEnabledSubject.next(enabled);
    }

    get equippedMusic(): string {
        return this._equippedMusic;
    }

    playBackgroundMusic(filename: string = this._equippedMusic): void {
        // Stop and recreate player if switching to a different track
        if (this.musicPlayer && this.currentlyLoadedMusic !== filename) {
            this.stopMusic();
            this.musicPlayer = null;
            this.currentlyLoadedMusic = null;
        }

        if (!this.musicPlayer) {
            this.musicPlayer = new Audio(`assets/sounds/${filename}`);
            this.musicPlayer.loop = true;
            this.musicPlayer.volume = 0.5;
            this.currentlyLoadedMusic = filename;
        }

        if (this.isMusicEnabled) {
            this.musicPlayer.play().catch((error) => {
                console.error('Error playing music:', error);
            });
        }
    }

    setEquippedMusic(filename: string): void {
        const wasPlaying = this.musicPlayer && !this.musicPlayer.paused;
        const wasMusicEnabled = this.isMusicEnabled;
        this._equippedMusic = filename;

        if (wasPlaying) {
            // Stop the current player without changing the enabled state
            if (this.musicPlayer) {
                this.musicPlayer.pause();
                this.musicPlayer.currentTime = 0;
                this.currentlyLoadedMusic = null;
            }
            this.musicPlayer = null;
            // Only play if music was enabled
            if (wasMusicEnabled) {
                this.playBackgroundMusic(filename);
            }
        }
    }

    pauseMusic(): void {
        if (this.musicPlayer && !this.musicPlayer.paused) {
            this.musicPlayer.pause();
        }
    }

    resumeMusic(): void {
        if (this.musicPlayer?.paused) {
            this.musicPlayer.play().catch((error) => {
                console.error('Error resuming music:', error);
            });
        }
    }

    stopMusic(): void {
        if (this.musicPlayer) {
            this.musicPlayer.pause();
            this.musicPlayer.currentTime = 0;
            this.currentlyLoadedMusic = null;
        }
        this.isMusicEnabledSubject.next(false);
    }

    playSoundEffect(filename: string, volume: number = 1): void {
        if (!this.areSoundEffectsEnabled) return;
        const sound = new Audio(`assets/sounds/${filename}`);
        sound.volume = volume;
        sound.play().catch((error) => {
            console.error('Error playing sound effect:', error);
        });
    }

    get isHostControlled(): boolean {
        return this.isHostControlledSubject.value;
    }

    setHostControlledSettings(musicEnabled: boolean, sfxEnabled: boolean): void {
        this.isHostControlledSubject.next(true);
        this.musicEnabled = musicEnabled;
        this.areSoundEffectsEnabled = sfxEnabled;
    }

    clearHostControl(): void {
        this.isHostControlledSubject.next(false);
    }
}
